keps/sig-scheduling/20200116-coscheduling-plugin-based-on-scheduler-framework.md

```diff
@@ -0,0 +1,216 @@
+---
+title: Coscheduling plugin based on scheduler framework
+authors:
+  - "@denkensk"
+owning-sig: sig-scheduling
+reviewers:
+  - "@Huang-Wei"
+  - "@resouer"
+  - "@hex108"
+approvers:
+  - "@Huang-Wei"
+creation-date: 2020-01-16
+last-updated: 2020-01-16
+status: provisional
+---
+
+# Coscheduling plugin based on scheduler framework
+
+## Table of Contents
+
+<!-- toc -->
+- [Motivation](#motivation)
+- [Goals](#goals)
+- [Non-Goals](#non-goals)
+- [Use Cases](#use-cases)
+- [Terms](#terms)
+- [Proposal](#proposal)
+- [Design Details](#design-details)
+  - [PodGroup](#podgroup)
+  - [Coscheduling](#coscheduling)
+  - [Extension points](#extension-points)
+    - [QueueSort](#queuesort)
+    - [Pre-Filter](#pre-filter)
+    - [Permit](#permit)
+    - [UnReserve](#unreserve)
+- [Alternatives considered](#alternatives-considered)
+- [Graduation Criteria](#graduation-criteria)
+- [Testing Plan](#testing-plan)
+- [Implementation History](#implementation-history)
+- [References](#references)
+<!-- /toc -->
+
+## Motivation
+Kubernetes has become a popular solution for orchestrating containerized workloads. Due to limitation of Kubernetes scheduler, some offline workloads(ML/DL)  are managed in a different system. For the sake of improving cluster utilization and operation efficiency., we'd like to treat Kubneretes as a unified management platform. But the ML jobs have the characteristics of All-or-Nothing in the scheduling process, which requires all tasks of a job to be scheduled at the same time. If the job only start part of tasks, it will wait for other tasks be ready to begin to work. In worst case, all jobs are pending like deadlock. To solve the above problem, co-scheduling is needed for the scheduler. The new scheduler framework makes the goal possible.
+ 
+## Goals
+ Use scheduler plugin, which is the most Kubernetes native way, to implement coscheduling.
+ 
+## Non-Goals
+ Discuss the API definition of `PodGroup`.
+ 
+## Use Cases
+When running a Tensorflow/MPI job, all tasks of a job must be start together; otherwise, did not start anyone of tasks. If the job only start part of tasks, it will wait for other tasks be ready to start work. In worst case, all jobs are pending like deadlock.
```

Comment:
```suggestion
When running a Tensorflow/MPI job, all tasks of a job must start or none. If the job only start part of tasks, it will wait for other tasks to be ready to start work. In worst case, all jobs are pending leading to a deadlock.
```

Reply:
when you say "it will wait for other tasks", who exactly waits? the tasks themselves keep waiting? or are you talking conceptually that the tasks should by some mechanism keep waiting (and so in our scheduler this is achieve by waiting at Permit)?

Reply:
The tasks themselves keep waiting, when the job only start part of tasks, Started tasks will for not started tasks to be ready to begin to work. This will lead to waste of resources. So the tasks should by some mechanism keep waiting in scheduler, so in our scheduler this is achieve by waiting at Permit.